[SPARK-15933][SQL][STREAMING] Refactored DF reader-writer to use readStream and writeStream for streaming DFs

tdas opened the pull request:
## What changes were proposed in this pull request?

Currently, the DataFrameReader/Writer has method that are needed for streaming and non-streaming DFs. This is quite awkward because each method in them through runtime exception for one case or the other. So rather having half the methods throw runtime exceptions, its just better to have a different reader/writer API for streams.
- [x] Python API!!
## How was this patch tested?

Existing unit tests + two sets of unit tests for DataFrameReader/Writer and DataStreamReader/Writer.

tdas committed:
Refactored DF reader-writer to use readStream and writeStream for streaming DFs

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala:
```diff
@@ -0,0 +1,288 @@
+import org.apache.spark.sql.execution.streaming.StreamingRelation
+import org.apache.spark.sql.types.StructType
+
+@Experimental
```
add docs.

SparkQA commented:
**[Test build #60452 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60452/consoleFull)** for PR 13653 at commit [`a59498b`](https://github.com/apache/spark/commit/a59498bbe86609bc206de9b229052f35071049cb).

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamReader.scala:
```diff
@@ -0,0 +1,288 @@
+   * @since 2.0.0
+   */
+  @Experimental
+  def load(): DataFrame = {
```
@marmbrus @rxin Should this be `load()`? In the general case, 
```
sdf.readStream
    .format("myFormat")
    .load("stringIdentifier")
```

marmbrus replied:
Seems best to keep it consistent.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:
```diff
@@ -0,0 +1,401 @@
+   *
+   * Scala Example:
+   * {{{
+   *   df.write.trigger(ProcessingTime("10 seconds"))
```
update these examples.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:
```diff
@@ -0,0 +1,401 @@
+   * Saves the content of the streaming [[DataFrame]] in Parquet format at the specified path.
+   * This is equivalent to:
+   * {{{
+   *   format("parquet").save(path)
```
update `save` if this is changed.

marmbrus replied:
this should probably start with `df.writeStream.`

tdas replied:
I am removing the parquet() method for now. Its nice to have a single method that starts the query - start()

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:
```diff
@@ -0,0 +1,401 @@
+   * @since 2.0.0
+   */
+  @Experimental
+  def save(path: String): ContinuousQuery = {
```
@marmbrus @rxin @mateiz Should this be `save()`? The general case will be

```
val query = sdf
  .writeStream
  .format("myFormat")
  .save("stringIdentifier")
```

marmbrus replied:
Here I did like `start()` since its much different than a single `save()` operation (i.e. its async and keeps going).

tdas replied:
Yes. I agree too. But then by that logic. Should `parquet()` automatically start the query?

marmbrus replied:
`writeStream.parquet()` doesn't seem incongruous with its behavior to me.

tdas replied:
I think I will remove parquet for now.

tdas commented on sql/core/src/main/scala/org/apache/spark/sql/streaming/DataStreamWriter.scala:
```diff
@@ -0,0 +1,401 @@
+   * @since 2.0.0
+   */
+  @Experimental
+  def outputMode(outputMode: String): DataStreamWriter[T] = {
```
should this be shortened to just `mode`?

marmbrus replied:
Again, I'd probably go for consistency unless we think that we will need save modes in streaming (i'm not really sure how that would work).

tdas commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/test/DataStreamReaderWriterSuite.scala:
```diff
@@ -371,76 +384,12 @@ class DataFrameReaderWriterSuite extends StreamTest with BeforeAndAfter {
 
   private def newTextInput = Utils.createTempDir(namePrefix = "text").getCanonicalPath
 
-  test("check trigger() can only be called on continuous queries") {
```
Most of these tests are to check whether one method is called on the wrong type of DF or not. So all of these can be removed.

tdas commented on sql/core/src/test/scala/org/apache/spark/sql/test/DataFrameReaderWriterSuite.scala:
```diff
@@ -0,0 +1,231 @@
+}
+
+
+class DataFrameReaderWriterSuite extends QueryTest with SharedSQLContext {
```
These tests are very rudimentary set of tests to test the DataFrameReader code path. Note that in Spark 1.6, there were no tests at all. Also I believe that a lot of the DF functionality is tested through other test suites (e.g. partitioning columns is tested through PartitionedParquetSuite)

SparkQA commented:
**[Test build #60452 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60452/consoleFull)** for PR 13653 at commit [`a59498b`](https://github.com/apache/spark/commit/a59498bbe86609bc206de9b229052f35071049cb).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60452/
Test FAILed.

tdas committed:
Updated python API

SparkQA commented:
**[Test build #60460 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60460/consoleFull)** for PR 13653 at commit [`e118631`](https://github.com/apache/spark/commit/e118631fd5bbea962932a812addae21c89f6f4f2).

SparkQA commented:
**[Test build #60460 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60460/consoleFull)** for PR 13653 at commit [`e118631`](https://github.com/apache/spark/commit/e118631fd5bbea962932a812addae21c89f6f4f2).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DataStreamReader(object):`
  * `class DataStreamWriter(object):`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60460/
Test FAILed.

tdas committed:
Fixed python style

SparkQA commented:
**[Test build #60461 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60461/consoleFull)** for PR 13653 at commit [`bbfff70`](https://github.com/apache/spark/commit/bbfff70b0bac8b2c33ab1eb736405759bb391633).

SparkQA commented:
**[Test build #60461 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60461/consoleFull)** for PR 13653 at commit [`bbfff70`](https://github.com/apache/spark/commit/bbfff70b0bac8b2c33ab1eb736405759bb391633).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60461/
Test PASSed.

marmbrus commented:
Overall looks pretty good.  Feel free to merge after addressing comments / passing tests to avoid more conflicts.

tdas committed:
Merge remote-tracking branch 'apache-github/master' into SPARK-15933

tdas committed:
Addressed comments

tdas commented:
@marmbrus Overall, I have changed the following. 
- Renamed writeStream.save() to writeStream.start()
- Refactored writeStream.foreach() to not start the query. instead writeStream.foreach().start() starts the query
- Removed writeStream.parquet(path) for now. Its good to have a single method `start` actually start the background query. experience says that its much easier read code and debug.

SparkQA commented:
**[Test build #60516 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60516/consoleFull)** for PR 13653 at commit [`536c25e`](https://github.com/apache/spark/commit/536c25e00b154a467279ab0055ae98002bd70847).

SparkQA commented:
**[Test build #60516 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60516/consoleFull)** for PR 13653 at commit [`536c25e`](https://github.com/apache/spark/commit/536c25e00b154a467279ab0055ae98002bd70847).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60516/
Test FAILed.

tdas committed:
Fixed python style

SparkQA commented:
**[Test build #60519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60519/consoleFull)** for PR 13653 at commit [`29ca23b`](https://github.com/apache/spark/commit/29ca23b74c16440dbb1e440423615357ec8015bf).

SparkQA commented:
**[Test build #60519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60519/consoleFull)** for PR 13653 at commit [`29ca23b`](https://github.com/apache/spark/commit/29ca23b74c16440dbb1e440423615357ec8015bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60519/
Test PASSed.

zsxwing commented on python/pyspark/sql/readwriter.py:
```diff
@@ -905,6 +764,503 @@ def jdbc(self, url, table, mode=None, properties=None):
         self._jwrite.mode(mode).jdbc(url, table, jprop)
 
 
+class DataStreamReader(object):
```
nit: add new classes to `__all__ = ["DataFrameReader", "DataFrameWriter"]`

tdas replied:
I will fix this in the follow up pr #13673

tdas replied:
Done in #13673

zsxwing commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala:
```diff
@@ -40,6 +40,8 @@ class StreamingAggregationSuite extends StreamTest with BeforeAndAfterAll {
 
   import testImplicits._
 
+
```
nit: redundant spaces.

tdas replied:
Done in #13673

zsxwing commented on python/pyspark/sql/readwriter.py:
```diff
@@ -905,6 +764,503 @@ def jdbc(self, url, table, mode=None, properties=None):
+            raise TypeError("path can be only a single string")
+
+    @since(2.0)
+    def csv(self, path, schema=None, sep=None, encoding=None, quote=None, escape=None,
```
I'm wondering if we can create util methods for `csv` and `json` to make DataFrameReader and DataStreamReader share most of codes here. Otherwise, it's hard to make them consistent in future.

tdas replied:
I would like to do that in that in a future PR. I would like to get this in as fast as possible.

tdas commented:
I am merging this for now. Please keep reviewing this and I will update them in #13673
